File: CSharp/CSharp_S3AP_LW3E2/XYArrays.cs
//namespace CSharp_S3AP_LW3E1;

public class XYArrays{
    double[] x_array;
    double[] y_array;

    public XYArrays(){
        x_array = new double[10];
        for (int i = 0; i < x_array.Length; i++){
            x_array[i] = 1.0;
        }
        y_array = new double[10];
        for (int i = 0; i < y_array.Length; i++){
            y_array[i] = 1.0;
        }
    }

    public void makeX(){
        double temp = 1;
        for (int i = 0; i < x_array.Length; i++){
            x_array[i] = Math.Round(Math.Sin(i + 1) * temp, 3);
            temp = x_array[i];
        }
    }

    public void makeYfromX(){
        double sum = 0.0;
        double mean;
        double[] temp = new double[x_array.Length];
        int temp_counter = 0;
        double[] result = new double[x_array.Length];
        int result_counter = 0;
        foreach (double i in x_array){
            sum += i;
        }
        mean = sum / x_array.Length;
        foreach (double i in x_array){
            if (i < mean){
                result[result_counter++] = i;
            }
            else {
                temp[temp_counter++] = i;
            }
        }
        for (int i = 0; i < temp_counter &&
        result_counter < result.Length; i++){
            result[result_counter++] = temp[i];
        }
        setY(result);
    }

    public double[] getX(){
        return x_array;
    }

    public void setY(double[] y_array){
        this.y_array = y_array;
    }

    public void setX(double[] x_array){
        this.x_array = x_array;
    }

    public double[] getY(){
        return y_array;
    }
}

File: CPP/CPP_S3AP_LW3E1/xyarrays.cpp
#include "xyarrays.h"

XYArrays::XYArrays()
{
    this->x_array = std::vector<double>(N, 1.0);
    this->y_array = std::vector<double>(N, 1.0);
}

void XYArrays::makeX(){
    this->x_array[0] = std::sin(1);
    double temp = x_array[0];
    for (std::size_t i = 1; i < x_array.size(); i++){
        x_array[i] = std::sin(i+1) * temp;
        temp = x_array[i];
    }
}

void XYArrays::makeYfromX(){
    double sum = 0.0;
    double mean;
    std::vector<double> temp, result;
    for (auto el : this->x_array){
        sum += el;
    }
    mean = sum / this->x_array.size();
    for (auto el : this->x_array){
        if (el < mean){
            result.push_back(el);
        }
        else{
            temp.push_back(el);
        }
    }
    result.insert(result.end(), temp.begin(), temp.end());
    setY(result);
}

std::vector<double> XYArrays::getX(){
    return this->x_array;
}

void XYArrays::setY(const std::vector<double> &y_array){
    this->y_array = y_array;
}

void XYArrays::setX(const std::vector<double> &x_array){
    this->x_array = x_array;
}

std::vector<double> XYArrays::getY(){
    return this->y_array;
}
